typing/ctype.ml

```diff
@@ -3156,40 +3158,41 @@ and unify_row_field env fixed1 fixed2 rm1 rm2 l f1 f2 =
       if_not_fixed second (fun () -> link_row_field_ext ~inside:f2 f1)
   | _ -> raise_unexplained_for Unify
 
-let unify env ty1 ty2 =
+let unify uenv ty1 ty2 =
   let snap = Btype.snapshot () in
   try
-    unify env ty1 ty2
+    unify uenv ty1 ty2
   with
     Unify_trace trace ->
       undo_compress snap;
-      raise (Unify (expand_to_unification_error !env trace))
+      raise (Unify (expand_to_unification_error (get_env uenv) trace))
 
 let unify_gadt ~equations_level:lev ~allow_recursive_equations
       (env:Env.t ref) ty1 ty2 =
-  try
-    univar_pairs := [];
-    gadt_equations_level := Some lev;
-    let equated_types = TypePairs.create 0 in
-    set_mode_pattern ~allow_recursive_equations ~equated_types
-      (fun () -> unify env ty1 ty2);
-    gadt_equations_level := None;
-    TypePairs.clear unify_eq_set;
-    equated_types
-  with e ->
-    gadt_equations_level := None;
-    TypePairs.clear unify_eq_set;
-    raise e
-
-let unify_var env t1 t2 =
+  univar_pairs := [];
```

Comment:
Would be nice to remove this global state, too, but I suppose not in this patch.